### PR TITLE
AzureMonitor: change clashing built-in Grafana variables/macro names for Azure Logs

### LINKED ISSUE #16365: Grafana uses same internal variables as plugin Azure Monitor
Grafana has two built in time range variables in $__from and $__to. They are currently always interpolated as epoch milliseconds. These variables are only available in Grafana v6.0 and above.

http://docs.grafana.org/features/datasources/azuremonitor/#azure-log-analytics-builtin-variables

### PR DESCRIPTION
**Pull request by shavonn**
Fixes #16365

### CODE REVIEW

**Comment by shavonn** [comment]
Updated help with new var names. Updated conversion of epoch to ISO string.
```
Usage
| where TimeGenerated >= $__from_az and TimeGenerated <= $__to_az
| summarize count() by Computer, bin(TimeGenerated, $__interval)
| order by TimeGenerated asc
```

Panel JSON:
```JSON
{
  "aliasColors": {},
  "bars": false,
  "dashLength": 10,
  "dashes": false,
  "fill": 1,
  "gridPos": {
    "h": 9,
    "w": 12,
    "x": 12,
    "y": 9
  },
  "id": 3,
  "legend": {
    "avg": false,
    "current": false,
    "max": false,
    "min": false,
    "show": true,
    "total": false,
    "values": false
  },
  "lines": true,
  "linewidth": 1,
  "links": [],
  "nullPointMode": "null",
  "percentage": false,
  "pointradius": 2,
  "points": false,
  "renderer": "flot",
  "seriesOverrides": [],
  "spaceLength": 10,
  "stack": false,
  "steppedLine": false,
  "targets": [
    {
      "appInsights": {
        "groupBy": "none",
        "metricName": "select",
        "rawQuery": false,
        "rawQueryString": "",
        "spliton": "",
        "timeGrainType": "auto",
        "xaxis": "timestamp",
        "yaxis": ""
      },
      "azureLogAnalytics": {
        "query": "Usage\n| where TimeGenerated >= $__from_az and TimeGenerated <= $__to_az\n| summarize count() by Computer, bin(TimeGenerated, $__interval)\n| order by TimeGenerated asc",
        "resultFormat": "time_series",
        "workspace": "a2c1b44e-3e57-4410-b027-6cc0ae6dee67"
      },
      "azureMonitor": {
        "aggOptions": [
          "Total"
        ],
        "aggregation": "Total",
        "dimension": null,
        "dimensionFilter": "*",
        "dimensions": [],
        "metricDefinition": "Microsoft.ContainerService/managedClusters",
        "metricName": "kube_node_status_allocatable_cpu_cores",
        "resourceGroup": "aks-us-east1-rg",
        "resourceName": "aks-us-east1",
        "timeGrain": "auto",
        "timeGrains": [
          {
            "text": "auto",
            "value": "auto"
          },
          {
            "text": "1 minute",
            "value": "PT1M"
          }
        ]
      },
      "queryType": "Azure Log Analytics",
      "refId": "A",
      "subscription": "44693801-6ee6-49de-9b2d-9106972f9572"
    }
  ],
  "thresholds": [],
  "timeFrom": null,
  "timeRegions": [],
  "timeShift": null,
  "title": "Azure Available Cores",
  "tooltip": {
    "shared": true,
    "sort": 0,
    "value_type": "individual"
  },
  "type": "graph",
  "xaxis": {
    "buckets": null,
    "mode": "time",
    "name": null,
    "show": true,
    "values": []
  },
  "yaxes": [
    {
      "format": "short",
      "label": null,
      "logBase": 1,
      "max": null,
      "min": null,
      "show": true
    },
    {
      "format": "short",
      "label": null,
      "logBase": 1,
      "max": null,
      "min": null,
      "show": true
    }
  ],
  "yaxis": {
    "align": false,
    "alignLevel": null
  }
}
```

**Review by daniellee — CHANGES_REQUESTED** [review]
This is mostly fine but there are a couple of improvements you could make.

- From a user's perspective the variable name `$__from_az` is a bit weird. Marcus suggested `$__rangeFrom` and `$__rangeTo` instead. 
- There is no typeahead/autocomplete in Slate - this is maybe too involved for this PR. Can add it in a separate PR (I can show you the Slate code tomorrow).
- The documentation needs to be updated.

**Comment by torkelo** [comment]
Is the az or azure suffix not a good idea to not confuse them with the built in variables?

**Comment by daniellee** [comment]
@torkelo good point but seems just weird to me. The other macros don’t have it. Looking at other datasources, like MySQL they don’t have anything like that.

On that point, saw that the sql datasources have timeFrom and timeTo which is both a better name and already a standard macro name for this concept.

**Comment by torkelo** [comment]
Oh, I see. Maybe it would be good to use the same names as the sql datasources then

**Review by daniellee — CHANGES_REQUESTED** [review]
Instead of a field like: `$__timeFrom` can you change it be a function: `$__timeFrom()`. Just to keep it consistent with the other datasources.

If you do that, then it is also easy to add typeahead for the two macros. Just add two entries to this array:

https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/grafana-azure-monitor-datasource/editor/kusto/kusto.ts#L646

**Comment by daniellee** [comment]
Also, docs need to be updated, as this section will now be incorrect: https://grafana.com/docs/features/datasources/azuremonitor/#azure-log-analytics-builtin-variables

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/log_analytics/querystring_builder.ts [hunk 1]
```diff
@@ -21,12 +21,18 @@ export default class LogAnalyticsQuerystringBuilder {
...
+        if (p1 === 'timeTo') {
+          return this.getUntil(this.options);
+        }
 
         return match;
       });
       queryString = queryString.replace(/\$__interval/gi, this.options.interval);
-      queryString = queryString.replace(/\$__from/gi, this.getFrom(this.options));
-      queryString = queryString.replace(/\$__to/gi, this.getUntil(this.options));
+      // queryString = queryString.replace(/\$__timeFrom/gi, this.getFrom(this.options));
```
Comment (shavonn) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/log_analytics/querystring_builder.ts]: @daniellee should I leave support for the variables or just remove?

Comment (torkelo) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/log_analytics/querystring_builder.ts]: I see no point in keeping them

**Comment by shavonn** [comment]
> Also, docs need to be updated, as this section will now be incorrect: https://grafana.com/docs/features/datasources/azuremonitor/#azure-log-analytics-builtin-variables

Should be updated (assuming I got the right place).

**Review by daniellee — CHANGES_REQUESTED** [review]
@shavonn one last thing - the migration tries to migrate `$__timeFrom` and `$__timeTo` to `$__timeFrom()` and `$__timeTo()` but no users have `$__timeFrom`/`$__timeTo` in their queries. They have the $__from and $__to so the migration needs to replace those.

**Review by daniellee — APPROVED** [review]

**Pull request merged**